### PR TITLE
add sys prompt to config

### PR DESCRIPTION
**Pull request by reidliu41**
```
$ ilab model chat --help
Usage: ilab model chat [OPTIONS] [QUESTION]...

  Runs a chat using the modified model

Options:
  --system-prompt TEXT            Custom system prompt. If not provided, it
                                  will use the default context. [default:
                                  <None>; config: 'chat.system_prompt']
```

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if nessessary.
- [ ] E2E Workflow tests have been added, if necessary.

Easy updates: can change the system prompt without touching the code.
Customization: Users can set their own prompt to fit different needs.
Consistency: Ensures the same prompt is used everywhere.
Flexibility: Prepares for future use of multiple prompts.
Better organization: Keeps settings separate from the code.

### CODE REVIEW

File: src/instructlab/configuration.py [hunk 1]
```diff
@@ -111,6 +111,13 @@ class _InstructlabDefaults:
         "lora_target_modules": ["q_proj", "k_proj", "v_proj", "o_proj"],
     }
 
+    # System Prompt
+    SYS_PROMPT = (
```
Comment (leseb) [inline src/instructlab/configuration.py:115]: Why not re-use the prompt in src/instructlab/common.py?

File: src/instructlab/configuration.py [hunk 2]
```diff
@@ -661,6 +665,15 @@ class _train(BaseModel):
     )
 
 
+class _prompts(BaseModel):
```
Comment (cdoern) [inline src/instructlab/configuration.py:668]: is this used anywhere? I don't see this being use directly in a cmd so I am hesitant to add a field without the usecase

**Review by cdoern — CHANGES_REQUESTED** [review]

File: src/instructlab/configuration.py (hunk 2 above)
Comment (reidliu41) [reply src/instructlab/configuration.py:668]: Please correct me if I am misunderstanding, here is what I check:
1. Currently, the `SYS_PROMPT` is using in `chat` from the handcode `common.py`
2. Due to different models, or different situations for the user, so the prompt might be need to modify by user frequently.
3. Change from here: define a default one from `common.py`  -->     write into `config.yaml` with init --> read the prompt from the config for `chat` --> user can use `ilab config edit` to modify the prompt

Comment (nathan-weinberg) [reply src/instructlab/configuration.py]: @cdoern can you respond to the above please

Comment (reidliu41) [reply src/instructlab/configuration.py]: thanks for response.  I do see a user ask for change(system prompt) in `users` channel, may consider set it in config.

File: src/instructlab/configuration.py [hunk 3]
```diff
@@ -564,6 +578,11 @@ class Config(BaseModel):
         description="Configuration file structure version.",
         frozen=True,  # don't allow this to be changed anywhere in the code
     )
+    # prompts configuration
+    prompts: _prompts = Field(
```
Comment (alimaredia) [inline src/instructlab/configuration.py:582]: Since this field is only being used in chat, it should only be included in `_chat`.

File: src/instructlab/common.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
 # SPDX-License-Identifier: Apache-2.0
 
 SYS_PROMPT = "I am, Red Hat® Instruct Model based on Granite 7B, an AI language model developed by Red Hat and IBM Research, based on the Granite-7b-base language model. My primary function is to be a chat assistant."
+CLI_HELPER_PROMPT = "You are an expert for command line interface and know all common commands. Answer the command to execute as it without any explanation."
```
Comment (alimaredia) [inline src/instructlab/common.py:4]: What if we just added these prompts to where you're setting them in `defaults.py` and removed this file. I tried to look at the git history for this file and didn't find much.

Removing this file would mean a change would have to be made to `utils.py`.

File: src/instructlab/utils.py [hunk 5]
```diff
@@ -367,7 +367,9 @@ def get_sysprompt(model=None):
     Returns:
         str: The system prompt for the model being used
     """
-    return common.SYS_PROMPT
+    config = read_config()
```
Comment (alimaredia) [inline src/instructlab/utils.py:370]: It looks like you can get rid of the `model` argument here too, As of this change it doesn't look like we have plans to implement it.

**Review by alimaredia — CHANGES_REQUESTED** [review]

File: src/instructlab/utils.py (hunk 5 above)
Comment (reidliu41) [reply src/instructlab/utils.py:370]: hey thanks for your feedback. Seems I see @jaideepr97 working on another pr that related to this prompt? Should I continue on this?

Comment (alimaredia) [reply src/instructlab/utils.py:370]: Is this the PR you're mentioning: https://github.com/instructlab/instructlab/pull/2592? 

From what I can tell, it looks like you're trying to give the user the ability to set the system prompt if they wish, which I think is great, and I don't see that functionality in #2582.

If that's the case, see how he feels about the changes here, and possibly expect to rebase and merge this PR in after he's done with his.

Comment (reidliu41) [reply src/instructlab/utils.py:370]: > It looks like you can get rid of the `model` argument here too, As of this change it doesn't look like we have plans to implement it.

yeah, model put it in there for a long time

File: src/instructlab/common.py (hunk 4 above)
Comment (reidliu41) [reply src/instructlab/common.py:4]: yeah, should only use in chat so far. I think is ok.

Comment (reidliu41) [reply src/instructlab/common.py:4]: if just put it in `default.py`, seems no need to use `get_sysprompt `. Actually, seems even in common, seems also no need to use `get_sysprompt `, may be different purpose?

**Comment by jaideepr97** [comment]
@reidliu41 I think a left a comment about BYO sys prompt on a different PR
https://github.com/instructlab/instructlab/pull/2577#pullrequestreview-2424214066

TL;DR we may need to wait for some research input on what the consequences of allowing BYO sys prompt are and what the best way to expose this feature to users would be 

cc @Maxusmusti

**Review by jaideepr97 — COMMENTED** [review]
@reidliu41 I think a left a comment about BYO sys prompt on a different PR
https://github.com/instructlab/instructlab/pull/2577#pullrequestreview-2424214066

TL;DR we may need to wait for some research input on what the consequences of allowing BYO sys prompt are and what the best way to expose this feature to users would be

cc @Maxusmusti

File: tests/testdata/default_config.yaml [hunk 6]
```diff
@@ -19,6 +19,9 @@ chat:
   # Filepath of a dialog session file.
   # Default: None
   session:
+  # Custom system prompt. If not provided, it will use the default context.
+  # Default: ''
+  system_prompt: ''
```
Comment (RobotSail) [inline tests/testdata/default_config.yaml:24]: Do we really want the default system system prompt to be set to an empty string? It seems like a better default would be simply not including it / having it set to `None`.

Comment (alimaredia) [reply tests/testdata/default_config.yaml:24]: The default in pydantic is `None` (https://github.com/instructlab/instructlab/blob/main/src/instructlab/configuration.py#L527). Not sure why it's rendering to an empty string here.

Comment (reidliu41) [reply tests/testdata/default_config.yaml]: thanks, updated

File: tests/testdata/default_config.yaml [hunk 7]
```diff
@@ -19,6 +19,9 @@ chat:
   # Filepath of a dialog session file.
   # Default: None
   session:
+  # Custom system prompt. If not provided, it will use the default context.
+  # Default: None
+  system_prompt:
```
Comment (RobotSail) [inline tests/testdata/default_config.yaml:24]: Is there any way for it to not appear in the config at all? Or is this a limitation of the current config system?

**Review by RobotSail — COMMENTED** [review]
I think this is a good change, the only issues is that we have a bunch of instances where the system prompt is referenced throughout the codebase. I'm seeing that the amount of lines changed is only 23, are we sure that we're covering all of them? 

If not, I think it's fine to leave it as something that's specific to chat, but we should probably add some tests if possible to validate the behavior.

File: tests/testdata/default_config.yaml (hunk 7 above)
Comment (reidliu41) [reply tests/testdata/default_config.yaml:24]: I think so, seems  a limitation of the current config system. Tried `exclude=True`, still will appear..

**Comment by reidliu41** [comment]
```
    of.gguf.write_tensor_data(ndarray)
  File "/actions-runner/_work/instructlab/instructlab/venv/lib64/python3.11/site-packages/gguf/gguf_writer.py", line 417, in write_tensor_data
    tensor.tofile(fout)
OSError: Not enough free space to write 67108864 bytes
[ 89/291] Writing tensor blk.9.ffn_down.weight                  | size   4096 x  11008  | type F32  | T+  65
[ 90/291] Writing tensor blk.9.attn_norm.weight                 | size   4096           | type F32  | T+  66
```

**Comment by reidliu41** [comment]
> I think this is a good change, the only issues is that we have a bunch of instances where the system prompt is referenced throughout the codebase. I'm seeing that the amount of lines changed is only 23, are we sure that we're covering all of them?
> 
> If not, I think it's fine to leave it as something that's specific to chat, but we should probably add some tests if possible to validate the behavior.

yes, so far only use for chat. 
I reviewed the others, seems some use the same , some just similar(should be different), so now separate it , might be good...

File: tests/testdata/default_config.yaml (hunk 7 above)
Comment (reidliu41) [reply tests/testdata/default_config.yaml:24]: lol, yeah, but seems evaluate do it too:'
https://github.com/instructlab/instructlab/blob/5922a76266ad65b4cfc1bb7ffeee9de56dd96df6/tests/testdata/default_config.yaml#L112

Comment (reidliu41) [reply tests/testdata/default_config.yaml:24]: I saw @alimaredia submit a `redesign config`, maybe he has ideas about this.

Comment (RobotSail) [reply tests/testdata/default_config.yaml:24]: Yeah I suppose it's fine, it just looks very strange.

**Review by RobotSail — APPROVED** [review]
LGTM

**Comment by reidliu41** [comment]
@alimaredia hi, can you take a look?

**Review by booxter — APPROVED** [review]

**Comment by reidliu41** [comment]
hi team, can someone take a look?

**Comment by mairin** [comment]
@reidliu41 it looks like it's blocked because @cdoern and @alimaredia requested changes... are those still applicable?

**Comment by reidliu41** [comment]
thanks, yeah, so if someone has time, might help to review..

**Comment by courtneypacheco** [comment]
Hey @reidliu41!

A PR was opened recently that is similar to yours: https://github.com/instructlab/instructlab/pull/3197

The difference is that PR focuses on decoupling the notion of hardcoding the specific model for each different workflow (training, sdg, eval, etc). Can you take a look?

**Comment by reidliu41** [comment]
I think https://github.com/instructlab/instructlab/pull/3197 is not related to `system_prompt` in `chat`,  please correct me if I misunderstood, @RobotSail

**Comment by reidliu41** [comment]
@RobotSail thank you so much for your explanation. It's good to wait for https://github.com/instructlab/instructlab/pull/3197 finish.

**Comment by RobotSail** [comment]
@reidliu41 We should rebase onto the new main and update this PR

**Comment by reidliu41** [comment]
@RobotSail updated
